### ISSUE TITLE
minor speedup in pcg32_random_r() possible?

### ISSUE THREAD

**Issue by jason-s** [imneme/pcg-c-basic#3]
`return (xorshifted >> rot) | (xorshifted << ((-rot) & 31));`

Wouldn't this be equivalent (and hence be slightly faster)?

`return (xorshifted >> rot) | (xorshifted << (32-rot));`

**Comment by imneme** [comment]
No.  Not when rot == 0.  Your code would shift by 32, and that’s undefined behavior.

**Comment by imneme** [comment]
FWIW, when the compiler's idiom-detection is working properly, it should look at this code and say “Oh, that's an idiomatic C/C++ rotate, I'll compile that to a single rotate instruction”.  If the optimizer doesn't do that, it's not really doing its job and should be reported as a bug.

**Comment by jason-s** [comment]
ah -- [those darn standards writers](http://stackoverflow.com/questions/18799344/shifting-a-32-bit-integer-by-32-bits). Thanks! 

(editorial comment: Seems like the C standard library should have contained a rotate intrinsic, rather than forcing programmers to jump through tricky hoops to do what they really want.)

**Comment by imneme** [comment]
Agreed!

**Comment by travisdowns** [comment]
@jason-s - in this case I think the standard writers chose a reasonable tack: ISA do actually differ in what they do when do a shift by 32. Some make the result zero, x86 leaves the result unchanged (effectively, the shift amount is taken mod 32), ARM is a mix of both.
